### ISSUE TITLE
Potential bug in `ctx: {session?: SessionContext} = {}` and then its `ctx.session!` usage

### ISSUE THREAD

**Issue by kripod** [blitz-js/blitz#815]
### What is the problem?

Why is the following argument used for the context when the [session middleware is in place](https://blitzjs-com-git-auth.blitzjs.vercel.app/docs/auth#setup)?

```js
function someQuery(input: any, ctx: {session?: SessionContext} = {}) { /* … */ }
```

This holds the potential for unexpected errors when using like `ctx.session!`, as shown in the auth example docs. I would suggest the following syntax if the `session` is always available:

```js
function someQuery(input: any, ctx: {session: SessionContext}) { /* … */ }
```

Otherwise, if the `ctx` argument may be missing, then check for its existence in runtime instead of falsely asserting something with TypeScript.

**Comment by kripod** [comment]
I realized that the initial method signature is like that to make it callable without explicitly specifying a second argument:

> - `ctx: Record<any, unknown>`
> [This context object is automatically provided by Blitz at run time.](https://blitzjs-com-git-auth.blitzjs.vercel.app/docs/query-definition#arguments)

That didn't seem to be intuitive at first. I think a wrapper like `ssrQuery` should always be present even on the server, and for both queries and mutations.

**Comment by merelinguist** [comment]
cc @flybayer

**Comment by flybayer** [comment]
@kripod so sorry for the long delay responding! I was traveling this past week and focused on getting the auth release out the door.

You are totally right about this being weird. This is the one remaining sore spot in Blitz apps that we need to find a solution for.

We could definitely add a higher order function to solve this, but if possible I'd love to keep the default export for queries and mutations as a plain, unwrapped function. This needs some exploration and probably some discussion with TS experts.

If anyone has ideas on this, please share!

**Comment by flybayer** [comment]
The other thing here is how to add types for global middleware. Like if you add custom global middleware, what's the best way to get that fully typed in all your queries and mutations.

**Comment by flybayer** [comment]
[Here's an RFC](https://github.com/blitz-js/blitz/discussions/1052) for a solution to this!

**Comment by flybayer** [comment]
Closed in #1160
